staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go

```diff
@@ -378,6 +386,39 @@ func (s *APIAggregator) PrepareRun() (preparedAPIAggregator, error) {
 		})
 	}
 
+	s.discoveryManager = NewDiscoveryManager(
+		aggregatorscheme.Codecs,
+		s.GenericAPIServer.Serializer,
+		func(a string) http.Handler {
+			return s.proxyHandlers[a]
+		},
+	)
+
+	// Inform discovery manager of all local api servers which contain
+	// crds/builtin types
+	s.discoveryManager.AddLocalAPIService("kube-aggregator", s.GenericAPIServer.DiscoveryResourceManager)
+
+	i := 0
+	for delegate := s.GenericAPIServer.NextDelegate(); delegate != nil && delegate.NextDelegate() != nil; delegate = delegate.NextDelegate() {
+		s.discoveryManager.AddLocalAPIService("delegate_"+strconv.Itoa(i), delegate.UnprotectedHandler())
+		i++
+	}
+
+	// Setup discovery endpoint
+	s.GenericAPIServer.Handler.GoRestfulContainer.Add(s.discoveryManager.WebService())
+
+	s.GenericAPIServer.AddPostStartHookOrDie("apiservice-discovery-initial", func(context genericapiserver.PostStartHookContext) error {
+		// Call synchronous function to setup initial document once server starts
+		return s.discoveryManager.RefreshDocument()
```

Comment:
This function waits for ALL discovery documents to refresh (not just crd/builtins). I think that is not desired